[2201.10.0] Update release note with OpenAPI changes (#9231)
ballerina-platform/ballerina-dev-website: downloads/swan-lake-release-notes/swan-lake-2201.10.0/RELEASE_NOTE.md

    ```

#### OpenAPI tool

> -  Added support to generate a mock client for OpenAPI specifications (OAS) that include examples. To generate a mock client, use the `--mock` flag in the OpenAPI client generation CLI command.
>   ```
>   $ bal openapi -i <yml file> --mode client --mock
>   ```

Reviewer: ````suggestion

-  Added support to generate a mock client for OpenAPI specifications (OAS) that include examples. To generate a mock client, use the `--mock` flag in the OpenAPI client generation CLI command.

  ```
  $ bal openapi -i <yml file> --mode client --mock
  ```

````

Please fix new lines for all the points.

Author: Sorry few comments getting conflicts while resolvings , will add

  ```
  $ bal openapi -i <yml file> --mode <client|service> --single-file
  ```
-  Added support to generate the Ballerina service contract object type for a given OAS. To generate the service contract object type, use the `--with-service-contract` flag along with the OpenAPI CLI client/service generation command.
  ```
  $ bal openapi -i <yml file> --mode <client|service> --with-service-contract
  ```
-  Introduced example annotations in the OpenAPI package. This feature will  enable rendering example schemas in the generated OpenAPI specification.
ex:
  1. Using the `openapi:Example` annotation
 ```ballerina
    @openapi:Example {
      value: {
        id: 10,
        name: "Jessica Smith"
      }
    }
    type User record {
      int id;
      string name
    }
 ```
  2. Using the `openapi:Examples` annotation
  ```ballerina
    @openapi:Examples {
      Jessica: { // Example 1
        value: {
           id: 10,
           name: "Jessica Smith"
        }
      },
      Ron: { // Example 2
        value: {
           id: 11,
           name: "Ron Stewart"
        }
      } 
    }
    type User record {
      int id;
      string name
    }
  ```

> - Provided a flag to generate the Ballerina client/service adhering to Ballerina naming conventions. To enable this feature, use the `--use-sanitized-oas` flag in the OpenAPI client/service CLI command. This an experimental feature.

Reviewer: Was `sanitized-oas` the correct term here? Aren't we sanitizing the generated code?

Author: Yeah, this flag modifies the given YAML according to Ballerina naming conventions. Currently, we generate the client/service based on the user-provided YAML without engaging any Ballerina naming conventions.

  ```
  $ bal openapi -i <yml file> --mode <client|service> --use-sanitized-oas
  ```
### Improvements

#### Language Server

- Added navigation and reference-finding support for object `init` functions.

> #### OpenAPI tool

Reviewer: ```suggestion
#### OpenAPI tool

```

> - Added support for Server-sent events (SSE) in the Ballerina client generation.

Reviewer: ```suggestion
- Added support for Server-sent events (SSE) in Ballerina client generation.
```

> - Added support for configuring the `examples` field in `@openapi:ResourceInfo` annotation using a reference path to the JSON example file.

Reviewer: ```suggestion
- Added support to specify the path to an example file in JSON format as the `examples` field in the `@openapi:ResourceInfo` annotation.
```

    ```

#### OpenAPI tool

> -  Added support to generate a mock client for OAS if including the example details. To generate mock client, use `--mock` flag along with openAPI CLI client generation command.

Reviewer 2: ```suggestion
-  Added support to generate a mock client from OAS which includes the example details. To generate the mock client, use `--mock` flag along with the OpenAPI CLI client generation command.
```

  ```
  $ bal openapi -i <yml file> --mode client --mock
  ```

> -  Provided the option to generate a single bal file for client/service. To generate single file, use `--single-file` flag along with openAPI CLI client/service generation command.

Reviewer 2: ```suggestion
-  Provided an option to generate a single ballerina file with the client or the service. To generate in a single file, use `--single-file` flag along with the OpenAPI CLI client/service generation command.
```

  ```

>   $ bal openapi -i <yml file> --mode <client/service> --single-file

Reviewer 2: ```suggestion
  $ bal openapi -i <yml file> --mode <client|service> --single-file
```

  ```

> -  Added support to generate Ballerina service object type from given OAS. To generate a service contract object type file, use the `--with-service-contract` flag along with the OpenAPI CLI service generation command.

Reviewer 2: ```suggestion
-  Added support to generate the Ballerina service contract object type for a given OAS. To generate the service contract object type, use the `--with-service-contract` flag along with the OpenAPI CLI client/service generation command.
```

Author: @MaryamZi , I'm a bit confused about using the keywords `flag` and `option`. If we require additional information or arguments to specify a value or a set of values, we use options. and if we only provide boolean switches that turn on or off a certain feature or behavior, we mention them as flags. In our use case, we have boolean switches rather than providing values. Please correct me if I have misunderstood this.

Reviewer: Yeah, I guess it's okay then. But, we generally stick to option. E.g., help text.

  ```

>   $ bal openapi -i <yml file> --mode service --with-service-contract

Reviewer 2: ```suggestion
  $ bal openapi -i <yml file> --mode <client|service> --with-service-contract
```

  ```

> -  Introduced example annotations in the OpenAPI tool. This feature will  enable rendering example schemas in the generated OpenAPI specification.

Reviewer 2: ```suggestion
-  Introduced example annotations in the OpenAPI package. This feature will  enable rendering example/examples schemas in the generated OpenAPI specification.
```

Reviewer: I think we can keep example.

> ex:

Reviewer: Not necessary.

Add new line.

Reviewer: Still present.

Reviewer: If you need to keep it has to be indented and 

For example,

>   1. Using Example annotation

Reviewer: ```suggestion
  1. Using the `openapi:Example` annotation
```

 ```ballerina
    @openapi:Example {
      value: {
        id: 10,
        name: "Jessica Smith"
      }
    }
    type User record {
      int id;
      string name
    }
 ```
  2. Using Examples annotation
  ```ballerina

>     @openapi:Examples {

Reviewer: Why were two different annotations introduced instead of one annotation that is allowed multiple times? Do the keys "Jessica", "Ron", etc. mean anything here?

Author: We introduced these two annotations, which represent the OAS properties themselves. In OAS, we have two different types of schema: `example` and `examples`. The keys have in relation to the OAS; if it is an `examples` schema, a key is always required for mapping it in the OAS schema.

Reviewer: Ack, thanks.

      Jessica: { // Example 1
        value: {
           id: 10,
           name: "Jessica Smith"
        }
      },
      Ron: { // Example 2
        value: {
           id: 11,
           name: "Ron Stewart"
        }
      } 
    }
    type User record {
      int id;
      string name
    }

>   ```

Reviewer: ````suggestion
  ```

````

> - Provided an option to generate Ballerina client/service according to Ballerina friendly naming convention. To enable this feature, use `--use-sanitized-oas` flag along with client/service CLI code generation.

Reviewer 2: ```suggestion
- Provided an option to generate the Ballerina client/service according to Ballerina friendly naming convention. To enable this feature, use `--use-sanitized-oas` flag along with the OpenAPI client/service CLI command. This an experimental feature.
```

Reviewer: ```suggestion
- Provided an option to generate the Ballerina client/service adhering to Ballerina naming conventions. To enable this feature, use the `--use-sanitized-oas` option in the OpenAPI client/service CLI command. This an experimental feature.
```

  ```
  $ bal openapi -i <yml file> --mode <client/service> --use-sanitized-oas
  ```
  
### Improvements

#### Language Server

- Added navigation and reference-finding support for object `init` functions.

#### OpenAPI tool

> - Added support for Server Sent Events (SSE) in Ballerina Client generation.

Reviewer 2: ```suggestion
- Added support for Server Sent Events (SSE) in the Ballerina client generation.
```

Reviewer: ```suggestion
- Added support for Server-sent events (SSE) in the Ballerina client generation.
```

> - Added improvement for supporting reference JSON example file path for the examples field in `@openapi:ResourceInfo` annotation.

Reviewer 2: ```suggestion
- Added support for configuring the `examples` field in `@openapi:ResourceInfo` annotation using a reference path to the JSON example file.
```

> - Added request body example mapping support for `@openapi:ResourceInfo` annotation in OAS generation.

Reviewer 2: ```suggestion
- Added support for the request body example mapping through the `@openapi:ResourceInfo` annotation.
```

Reviewer: ```suggestion
- Added support for request body example mapping via the `@openapi:ResourceInfo` annotation.
```

#### OpenAPI tool

-  Added support to generate a mock client for OpenAPI specifications (OAS) that include examples. To generate a mock client, use the `--mock` flag in the OpenAPI client generation CLI command.
  
    ```
     $ bal openapi -i <yml file> --mode client --mock
    ```

-  Provided an option to generate a single Ballerina file with the client or the service code. To generate code in a single file, use the `--single-file` flag in the OpenAPI client/service generation CLI command.
  
    ```
    $ bal openapi -i <yml file> --mode <client|service> --single-file
    ```

-  Added support to generate the Ballerina service contract object type for a given OAS. To generate the service contract object type, use the `--with-service-contract` flag along with the OpenAPI CLI client/service generation command.

    ```
    $ bal openapi -i <yml file> --mode <client|service> --with-service-contract
    ```

> -  Introduced example annotations in the OpenAPI package. This feature will  enable rendering example schemas in the generated OpenAPI specification.

Reviewer: ```suggestion
-  Introduced example annotations in the OpenAPI package. This feature will enable rendering example schemas in the generated OpenAPI specification.
```
